Add symantec distrust evaluation worker

jvehent opened the pull request:
This mostly works, with a few quirks I'm still ironing out.

It probably needs to be smarter about evaluating the paths, and making sure that a valid path is only valid if it maps to a trusted root. Which brings the question of which root do we trust? (probably mozilla's)

For now, testing `tesla.com` returns the following:
> "not impacted, but blacklisted certs found in path: /C=US/O=VeriSign, Inc./OU=VeriSign Trust Network/OU=(c) 2006 VeriSign, Inc. - For authorized use only/CN=VeriSign Class 3 Public Primary Certification Authority - G5"

jcjones commented on worker/symantecDistrust/symantecDistrust.go:
```diff
@@ -0,0 +1,175 @@
+	}
+	for _, parent := range paths.Parents {
+		var theirReason string
+		trust, theirReason = evalPaths(parent)
```
The nature of this is that `trust` will be set `true` or `false` depending on the final parent checked, which is not likely what we want. It seems like we want it gated: either it's `true` if `evalPaths()` ever returns true, or it's the pessimistic case of  `false` if `evalPaths()` ever returns false. I'm not sure the UI implications of these choices, but IMO it should latch.

jvehent commented:
How much detail do you want in the result? I could return the list of paths
that successes and failed, and why, just a Boolean, and anything in between.

On Feb 24, 2018 19:51, "J.C. Jones" <notifications@github.com> wrote:

> *@jcjones* requested changes on this pull request.
>
> Approach looks good, and the text seems alright to me too. It might be
> nice to give a link to the Wiki page
> <https://wiki.mozilla.org/CA/Upcoming_Distrust_Actions> in the error
> text, too.
> ------------------------------
>
> In worker/symantecDistrust/symantecDistrust.go
> <https://github.com/mozilla/tls-observatory/pull/299#discussion_r170437612>
> :
>
> > +	res <- worker.Result{
> +		Success:    true,
> +		WorkerName: workerName,
> +		Errors:     nil,
> +		Result:     outB,
> +	}
> +}
> +
> +func evalPaths(paths certificate.Paths) (trust bool, reason string) {
> +	blacklisted, name := evalBlacklist(paths.Cert.Hashes.SHA256)
> +	if blacklisted {
> +		return false, name
> +	}
> +	for _, parent := range paths.Parents {
> +		var theirReason string
> +		trust, theirReason = evalPaths(parent)
>
> The nature of this is that trust will be set true or false depending on
> the final parent checked, which is not likely what we want. It seems like
> we want it gated: either it's true if evalPaths() ever returns true, or
> it's the pessimistic case of false if evalPaths() ever returns false. I'm
> not sure the UI implications of these choices, but IMO it should latch.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/tls-observatory/pull/299#pullrequestreview-99128307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZXgRgIQGJdhBfUt31k6aMp3ryW1vKRks5tYK56gaJpZM4SR8-D>
> .
>

jcjones commented:
I think it'd be fine to just be boolean, where if any of the paths are affected, we give a link to the wiki page?

jvehent commented:
Well, I was curious about the details, so I added some :) New output is
```
* Symantec distrust: impacted
  - path uses a blacklisted cert: C=US, O=VeriSign, Inc., OU=VeriSign Trust Network, OU=(c) 2006 VeriSign, Inc. - For authorized use only, CN=VeriSign Class 3 Public Primary Certification Authority - G5 (id=147)
  - path uses a root not trusted by Mozilla: C=US, O=VeriSign, Inc., OU=Class 3 Public Primary Certification Authority (id=154)
  - path uses a root not trusted by Mozilla: C=US, O=VeriSign, Inc., OU=Class 3 Public Primary Certification Authority (id=157)
  - for details, read https://wiki.mozilla.org/CA/Upcoming_Distrust_Actions
```
I inversed some of the logic so trust became distrust, and added a truststore check to only return impacted when a path of a root trusted by Mozilla cannot be found. That should be close enough to what Firefox does.

Remains the question of the whitelist you mentioned in #298. What, if anything, should we do about this?

jvehent commented:
Google.com shows as impacted, probably because we don't apply the whitelist?
```
* Symantec distrust: impacted
  - path uses a root not trusted by Mozilla: C=US, O=Equifax, OU=Equifax Secure Certificate Authority (id=63)
  - path uses a blacklisted cert: C=US, O=GeoTrust Inc., CN=GeoTrust Global CA (id=66)
  - for details, read https://wiki.mozilla.org/CA/Upcoming_Distrust_Actions
```

jvehent commented:
Id like to implement exactly the same logic as Firefox, so the whitelist is
important.

How does it work? I imagine it overwrites the distrust decision, but how?

On Feb 25, 2018 8:52 PM, "J.C. Jones" <notifications@github.com> wrote:

> *@jcjones* approved this pull request.
>
> This looks good to me now.
>
> Regarding Google (and Apple, for that matter): Yeah, this is because of
> not applying the whitelist, as you hypothesized. Up to you whether that's
> an issue....
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/tls-observatory/pull/299#pullrequestreview-99168986>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZXgXPsOL_o_igNARFu2O2UXFB69xk2ks5tYg5dgaJpZM4SR8-D>
> .
>

jvehent commented:
@jcjones does this look correct for google.com ?

```
--- Trust paths ---
C=US, O=Google Inc, CN=*.google.com (id=395)
└──C=US, O=Google Inc, CN=Google Internet Authority G2 (id=396)
   ├──C=US, O=GeoTrust Inc., CN=GeoTrust Global CA (id=397)
   │  └──C=US, O=Equifax, OU=Equifax Secure Certificate Authority (id=63)
   └──C=US, O=GeoTrust Inc., CN=GeoTrust Global CA (id=66)

* Symantec distrust: not impacted, but found paths with distrusted certs
  - path uses a blacklisted cert: C=US, O=GeoTrust Inc., CN=GeoTrust Global CA (id=66)
  - whitelisted intermediate C=US, O=Google Inc, CN=Google Internet Authority G2 (id=396) override blacklisting of 66
  - path uses a root not trusted by Mozilla: C=US, O=Equifax, OU=Equifax Secure Certificate Authority (id=63)
  - whitelisted intermediate C=US, O=Google Inc, CN=Google Internet Authority G2 (id=396) override blacklisting of 397
  - for details, read https://wiki.mozilla.org/CA/Upcoming_Distrust_Actions
```
